Add private `_nested_from_values_and_offsets`

mikaylagawarecki opened the pull request:
Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #95518

mikaylagawarecki committed:
Add private _nested_from_values_and_offsets

[ghstack-poisoned]

mikaylagawarecki commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -370,6 +371,12 @@ def emit_view_lambda(f: NativeFunction, unpacked_bindings: List[Binding]) -> str
+        elif (arg == "jagged_offsets_") and arg_type == ConstRefCType(
+            BaseCType(tensorT)
+        ):
+            # Jagged offsets must be kept on the GPU, to prevent sync so cannot be passed
```
I think follow up work would be to move _nested_tensor_size to the GPU in order to prevent the sync mentioned here (?)

cpuhrsch replied:
I think we'd need to move _nested_tensor_offset onto GPU, but size and stride can stay on CPU.

mikaylagawarecki replied:
hm but i'm iterating over the values from jagged_offsets (on GPU) to fill sizes a CPU tensor here, would that cause a sync?

cpuhrsch replied:
Yep, it definitely would. Another data point that treating the memory on the GPU like a cached value with the ground truth being readily available on the CPu might be a useful construct.

mikaylagawarecki replied:
what does it mean for the metadata on the GPU to be a cached value with the ground truth being readily available on the CPU?

mikaylagawarecki replied:
Do you want me to implement this before landing this PR?

cpuhrsch replied:
No, it's ok if this incur a sync for now. But we should track and discuss this.

mikaylagawarecki replied:
Great, tracking here #95815

mikaylagawarecki committed:
Update on "Add private `_nested_from_values_and_offsets`"

Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add private `_nested_from_values_and_offsets`"

Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add private `_nested_from_values_and_offsets`"

Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add private `_nested_from_values_and_offsets`"

Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add private `_nested_from_values_and_offsets`"

Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add private `_nested_from_values_and_offsets`"

Add conversion function that takes [sum(*), D] values Tensor and offsets Tensor and returns a nested tensor that is a view of the values tensor




[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13497,6 +13497,19 @@
   python_module: nested
   variants: function
 
+- func: _nested_from_values_and_offsets(Tensor(a) self, Tensor jagged_offsets) -> Tensor(a)
```
Noob q are self and jagged_offsets always on the same device?

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -919,6 +919,41 @@ Tensor _nested_view_from_buffer(
+ * @return Returns a nested tensor that
+ * aliases the same storage as values
+ */
+Tensor _nested_from_values_and_offsets(
```
I think that this function should be moved to NestedTensorFactories and if it relies on any things in Nt math either move them to header or if we think they are utils move them to Ntutils

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -919,6 +919,41 @@ Tensor _nested_view_from_buffer(
+  TORCH_CHECK(values.is_contiguous(), "values must be contiguous.");
+  TORCH_CHECK(offsets.dim() == 1, "offsets must be a 1D tensor, got offsets.dim() ", offsets.dim(), ".");
+
+  TORCH_CHECK(offsets.select(0, 0).item<int64_t>() == 0,
```
Could be wrong and we may want to measure/ may not be worth. I have always found that if you are working with a cpu tensor and are going to access its elements it is always substantially faster going to the data pointer first than indexing. It maybe less safe however but might be worth it for speed

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -2423,6 +2444,21 @@ def grad_test_func(a, b, c):
+
+
+        def grad_test_func(values, offsets):
+            nt = torch.nested._from_values_and_offsets(values, offsets)
```
Extend this graph a little and do some operations on the nt


As well try a mutating op in place. I think we have relu_

If any fail I think that's still okay but we need to track them and Xfail

drisspg reviewed (commented):
Couple small things

jbschlosser commented on torch/nested/__init__.py:
```diff
@@ -57,6 +57,11 @@ def as_nested_tensor(
 
+_from_values_and_offsets = _add_docstr(
+    _nested._nested_from_values_and_offsets,
+    """
+    """,
```
fill these in or TODO?

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -919,6 +919,41 @@ Tensor _nested_view_from_buffer(
+    "First element of offsets must be equal to 0, got ",
+    offsets.select(0, 0).item<int64_t>(),
+    ".");
+  TORCH_CHECK(offsets.select(0, -1).item<int64_t>() == values.size(0),
```
for `torch.compile()` purposes, we may not want to do these checks with item calls. There might be more done here that is incompatible with symbolically propagating shapes, TBD but I'll look into it

mikaylagawarecki replied:
That makes sense, I will make this private function an unsafe constructor for now then
